File: src/VB.Forms/Form1.vb
Imports VB.Repository
Imports VB.Repository.VB.Repository.Interfaces
Imports Microsoft.Extensions.Logging
Imports System.Windows.Forms.VisualStyles.VisualStyleElement.StartPanel
Imports VB.Forms.Data.VB.Forms.Data

Public Class Form1

    Private ReadOnly _userRepository As IUsersRepository
    Private ReadOnly _logger As ILogger

    Public Sub New(userRepository As IUsersRepository, logger As ILogger(Of Form1))
        _userRepository = userRepository
        _logger = logger

        ' This call is required by the designer.
        InitializeComponent()

        ' Add any initialization after the InitializeComponent() call.
    End Sub

    Private Async Sub bnAddUser_Click(sender As Object, e As EventArgs) Handles bnAddUser.Click
        Await AddUser()
    End Sub

    Private Async Function AddUser() As Task
        Dim userName = tbUserName.Text
        If userName.Length > 0 Then
            If Await _userRepository.UserExists(userName) Then
                _logger.LogError($"User {userName} already exists")
                MessageBox.Show($"User {userName} already exists")
            Else
                Await _userRepository.Add(New User(userName))
                userBindingSource.DataSource = Await _userRepository.GetAll()
                _logger.LogInformation($"Added user {userName}")
                tbUserName.Clear()
                tbUserName.Focus()
            End If
        End If
    End Function

    Private Async Sub bnRefresh_Click(sender As Object, e As EventArgs) Handles bnRefresh.Click
        userBindingSource.DataSource = Await _userRepository.GetAll()
    End Sub

    Private Async Sub tbUserName_KeyDown(sender As Object, e As KeyEventArgs) Handles tbUserName.KeyDown
        If e.KeyCode = Keys.Enter Then
            Await AddUser()
        End If
    End Sub

    Private Async Sub dataGridView1_UserDeletingRow(sender As Object, e As DataGridViewRowCancelEventArgs) Handles dataGridView1.UserDeletingRow
        If e.Row?.DataBoundItem IsNot Nothing Then
            Dim item As User = CType(e.Row.DataBoundItem, User)
            Await _userRepository.Delete(item)
            _logger.LogWarning($"Deleted user {item.Name}")
        End If
    End Sub

End Class

File: src/CSharp.Forms/Form1.cs
using CSharp.Forms.Data.Models;
using CSharp.Repository.Interfaces;
using Microsoft.Extensions.Logging;

namespace CSharp.Forms
{
    public partial class Form1 : Form
    {

        private readonly IUsersRepository _dbContext;
        private readonly ILogger _logger;

        public Form1(IUsersRepository usersRepository, ILogger<Form1> logger)
        {
            _logger = logger;
            _dbContext = usersRepository;

            InitializeComponent();
        }

        private async void bnAddUser_Click(object sender, EventArgs e)
        {
            await AddUser();
        }

        private async Task AddUser()
        {
            string userName = tbUserName.Text;
            if (userName.Length > 0)
            {
                if (await _dbContext.UserExists(userName))
                {
                    _logger.LogError($"User {userName} already exists");
                    MessageBox.Show($"User {userName} already exists");
                }
                else
                {
                    await _dbContext.Add(new Data.Models.User(userName));
                    userBindingSource.DataSource = await _dbContext.GetAll();
                    _logger.LogInformation($"Added user {userName}");
                    tbUserName.Clear();
                    tbUserName.Focus();
                }
            }
        }

        private async void tbUserName_KeyDown(object sender, KeyEventArgs e)
        {
            if (e.KeyCode == Keys.Enter)
                await AddUser();

        }

        private async void dataGridView1_UserDeletingRow(object sender, DataGridViewRowCancelEventArgs e)
        {
            if (e.Row?.DataBoundItem != null)
            {
                User item = (User)e.Row.DataBoundItem;
                await _dbContext.Delete(item);
                _logger.LogWarning($"Deleted user {item.Name}");
            }
        }

        private async void bnRefresh_Click(object sender, EventArgs e)
        {
            userBindingSource.DataSource = await _dbContext.GetAll();
        }
    }
}
